### ISSUE TITLE
__ZTISt9exception in poppler

### ISSUE THREAD

**Issue by MikeHolman** [kripken/Massive#19]
In poppler, __ZTISt9exception is imported as var o=env.__ZTISt9exception|0, but it is defined as:

    function __ZTISt9exception() {
      Module['printErr']('missing function: _ZTISt9exception'); abort(-1);
    }

[Asm.js errata](https://etherpad.mozilla.org/ibPVJ81u35) states that values for such imports must initially be primitives, so I believe this is not valid asm.js code and should fail at link time.

**Comment by kripken** [comment]
cc @lukewagner

I'm not sure I understand the errata there. But I don't think that is or should be invalid asm.js: whatever the fii input is, it is either left alone so it must be a function, or it is coerced to an int or a float, in which case it could be anything, the coercion makes it safe. Similarly to when we receive a result from an ffi, we must coerce, and that makes it safe for the ffi to return a function or an object (of course, it's silly).

As to why that code exists in the benchmark, I suspect it was a bug in a previous version of emscripten, when I build poppler now I don't see it.

I can update the version on gh-pages if you want.

**Comment by lukewagner** [comment]
The issue here is https://bugzilla.mozilla.org/show_bug.cgi?id=1053544: if imports have user-visible coercions (valueOf() call when performing |0), then the re-execution that happens when link-time validation fails would be observable.

We also ran into this bug when we first landed and had to add this special case:
  http://hg.mozilla.org/mozilla-central/file/22a157f7feb7/js/src/asmjs/AsmJSLink.cpp#l110
to allow functions (but only if their toString is the native one!).  I'm sorry for noting it in the Errata, but I had assumed it would've been been fixed by now.  For now, Mike, I'd add a similar exception or (if the double-execution of user-defined coercions isn't an issue in your design, ignore the whole bullet in the errata).
